### ISSUE TITLE
Format numbers

### ISSUE THREAD

**Issue by uchu** [ggxchain/rfq-ui#45]

**Comment by AuroraLantean** [comment]
@uchu  @Warchant  @akorchyn  
After some research, for financial calculation it is recommended to use integers to do all calculation, then we round up to what we want:  https://www.pullrequest.com/blog/precision-matters-why-using-cents-instead-of-floating-point-for-transaction-amounts-is-crucial/

Define "dp" as decimal precision...
- Addition/Subtraction: a tokenA 2dp balance adds or subtracts a tokenB 3dp balance:  Convert them into integers using the bigger dp
E.g. 0.01+0.001 => (10 + 1)x1000 = 11 and dp of 3 => 0.011, and that is 3 dp
Result dp = the max dp between them

- Multiplication: a tokenA 2dp balance multiplies a tokenB 3dp balance: Convert them into integers using the bigger dp
E.g. 0.01*0.001 => (10x1)/(100 x1000) => 1x1 and dp of 5 => 0.00001 and that is 5 dp. 
The result dp = the sum of those dp

- Division: a tokenA 3dp balance is divided by a tokenB 2dp balance: Convert them into integers using the bigger dp
3dp / 2dp => 2dp( top and bottom both are multiplied by the bigger of 10**dp => 0 dp / 0 dp => dp1-dp2 )
E.g.  0.001/0.10 => 1/10 => 0.1
The result dp = 3-2 = 1

2dp / 3dp => 2dp( top and bottom both are multiplied by the bigger of 10**dp => 0 dp / 0 dp => dp1-dp2 )
E.g. 0.01/0.001 => 10/1 = 10
The result dp = 0

E.g. 0.01/100.000 => 10/100000 = 0.0001 
The result dp = ?? ... Division in math is always strange... I think we can decide how much dp we want after division

**Comment by uchu** [comment]
E.g. 0.01/100.000 => 10/100000 = 0.0001
Why it cannot be dp = 4 here?

Lets add max dp as a configurable variable. We can start with something like 6 max, and we can adjust it later in one place

**Comment by AuroraLantean** [comment]
Yes, 4dp... I was saying the dp from division is not predictable given input of dp1 and dp2

All practical calculations, we can all use BigNumber or similar libraries in JavaScript here and Rust in the GGXNode
Those dp rounding are only for frontend users to view.
Okay I will add a max dp

**Comment by AuroraLantean** [comment]
Regarding "harcoded map: token: price precision, amount precision":
- Amount precision has already be determined by each token/currency, except rare cases
E.g. BTC: 8 dp, ETH: 18 dp, USDT on Ethereum network: 6 dp
E.g. Exception: USD should only have 2 dp, but I see Binance and Coinmarketcap show it in 4 dp. I think we should do that too to keep the accuracy.

- Price precision is determined by the base token(like base currency in foreign exchange), and I suggest we round it up to ceiling in displayed view. Ceiling rounding means 0.051 is rounded to 0.06 in 2 dp

E.g. the price of BTC in USDT should be the same as USDT 6dp, because the customers have to pay in USDT, which is in 6dp
61158.69 USD per BTC /0.9996 USD per USDT = 61183,163265306 USDT/BTC ... We round it to ceiling 6dp: 61183,163266

E.g. the price of USDT in BTC should be the same as BTC 8dp
0.9996 USD per USDT/ 61158.69 USD per BTC = 0.9996/61158.69 BTC/USDT = 0.000016344 ... We round it to ceiling 8dp: 0.00001635 

For actual exchange calculation when we give customer tokens, we could 
- calculate the input amount from rounded prices
OR
- calculate all in BigNumber to keep accuracy. So the customer may find they occasionally get slight more than expected tokens after exchange
please advise. Thanks

**Comment by AuroraLantean** [comment]
@uchu Regarding price and balance representation, after looking at Binance and Coinmarketcap, most tokens have 2 and 4 decimals, but the others have the following long decimals... However, they all have 4 significant numbers(DOGE changes in price a lot so it changes between 4 and 5 significant numbers). So 4 significant numbers is more important, not fixed decimals. Please advise. Thanks
Binance Price Display:
https://www.binance.com/en/markets/overview
DOGE    $0.15273 ... 5 dp
XME     $0.03685 ... 5dp
DGB     $0.01275 ... 5dp
XVG     $0.005628 ... 6dp
Coinmarketcap Price Display:
https://coinmarketcap.com/
DOGE    $0.1531 ... 4dp
XEM     $0.03696  ... 5 dp
VeChain $0.04085 ... 5 dp
Floki   $0.0001425 ... 7 dp
Shiba   $0.00002278 ... 8 dp
PEPE    $0.000005191 ... 9 dp
BONK    $0.00001483 ... 8 dp

**Comment by AuroraLantean** [comment]
@uchu From the above observation, Price Representation Rules can be deduced:
- DAI price uses 8 dp
- if price >= 1, use default 2 dp. This includes USDT and USDC: 1.00xy => 1.00, OR use 4 dp 0.999x
- if price < 1, use 4 significant figures(except 5 for DOGE, TRX)

**Comment by AuroraLantean** [comment]
I found another bug so cannot further progress on this item until this is fixed: https://github.com/orgs/ggxchain/projects/6/views/1?pane=issue&itemId=60517298

**Comment by akorchyn** [comment]
To be honest, this task seems related to this one. I don't think we need to create a new one and fix it as part of this one ☺️

**Comment by AuroraLantean** [comment]
@akorchyn the calculated price from `amountConverter.divWithPrecision(requested, offered,)`  is zero, so I will need to fix that whole calculation later on, and that is outside of formatting and rounding...

**Comment by AuroraLantean** [comment]
PR: "fix input amount decimals, balance decimals, price representation #54" https://github.com/ggxchain/rfq-ui/pull/54 ?
This PR includes all the following:
- fix price representation + decimal points + significant figures
- rename "consts.ts" to "settings.ts" according to Pavel
- fix tokenList unit test
- merge from main
`formatting-numbersPart2` branch is moved to `formatting-numbersPart3`
![1WalletPage](https://github.com/ggxchain/rfq-ui/assets/73493219/fe6caf0b-5350-413d-b166-c2b2de83aaab)

**Comment by AuroraLantean** [comment]
@uchu can I close this issue?

**Comment by uchu** [comment]
@AuroraLantean No, leave it to me please. I will review during the day. You can start with other tasks, that are some bugs. Thanks!

**Comment by AuroraLantean** [comment]
DexPage Orderlist Prices show zeros#272
And
DexPage SellAmount cannot handle some decimal inputs#268
Are done and waiting to be merged

**Comment by AuroraLantean** [comment]
I think all the issues regarding this item have been resolved and all the above PRs have been merged.
